### DISCUSSION TITLE
About thread assignment value and concurrent thread ability

### DISCUSSION THREAD

**Discussion by ghost** [nextgenhealthcare/connect#5698] [Q&A]
Hello everyone,

In my company, we are trying to use the multi-thread with thread assignment value. The value is actually the visit id of the patient. We've set 8 threads in the destination connector. What we were expected is that if one message is queued, Mirth would be able to handle message concerning other visit (as the thread assignment is again, the visit ID); but it seems that's is not actually the case. Example:

- Visit id 1 => Queued (by code in javascript, using `Responsefactory.getQueuedResponse("hey you made a mistake, correct it");` in destination connector); in the dashboard we can see a  Response Date.
- Visit id 2 => Sent; no problem
- Visit id 3 => Queued without any reason... Furthermore, we don't have any Response Date in the dashboard.

What we suspect the most is that it comes from the fact that server has only 2 CPU and is only able to handle 2 concurrent thread (even if the parameter is set to 8).

Does someone else encounter the same problem? Did we missed something?

Thanks a lot in advance for you help.

**Comment by jonbartels** [comment]
Can you show your thread-assignment-variable (TAV) setting exactly please? A common mistake is that some Mirth fields want a variable name, others want the Velocity `${myVar}` syntax. 

Are you using "Queue Always" or "Queue on Failure"?

What destination type? "Channel Writer" has some special behaviors you'll need to know about.

The hashing algorithm used should mostly distribute the TAV evenly between the threads. 

Modern CPUs will run multiple threads per core. 

Here is an old thread ported from the old JIRA where grandmaster Nick and some other OGs explain it https://github.com/nextgenhealthcare/connect/issues/3669

For "Queued without any reason"
* In the message dashboard, right-click on the header, and ensure the "send attempts" column is visible. How many send attempts for that queued message?
* In the message dashboard, view the queued message and see if the errors tab has any data. If a message errors, the errors tab will show the error for the most recent attempt, then re-queue the message.

**Reply by ghost** [reply]
So here the setup:
![image](https://user-images.githubusercontent.com/4374227/221257675-3c926a06-044b-40db-b522-2531cd450506.png)

The destination type is actually "Database Writer" with "Use Javascript => On"

As you can see, we're are using "Queue on Failure". We tried "Always" before and the result was actually worst. I just check the "send attemp" column and it says "0" ; no errors at all.
![image](https://user-images.githubusercontent.com/4374227/221259543-f8c10fde-3dd4-407d-b984-fc6477a3d7cb.png)
(It isn't show on the screenshot but you have a third line with a send attemp that is not 0 juste above the first one and that's it).

**Reply by jonbartels** [reply]
Relevant documentation:
* https://docs.nextgen.com/bundle/Mirth_User_Guide_40/page/connect/connect/topics/c_Advanced_Queue_Settingsconnect_ug.html
* https://docs.nextgen.com/bundle/Mirth_User_Guide_40/page/connect/connect/topics/c_Destination_Settings_connect_ug.html

Hmm your results are interesting. Two messages with a high count of send attempts does suggest a problem with a blocked queue. One message is trying to send over and over. With rotate queue off that message will try to send before moving on to the next one in the queue thread. 

Is there anything in the "Errors" tab for the messages with many send attempts? 

IMO if you don't ever need the response then "Queue Always" provides faster and more consistent behavior.

**Reply by ghost** [reply]
In fact, the code in the destination Javascript is something like this:
```javascript
if(!patientExist)
{
    return ResponseFactory.getQueuedResponse("Waiting for patient created");
}
else
{
    //Do plenty of cool stuff
}
```

So, we check that the patient is already created (this is handled by another channel) and if not, we want to wait and retry. We intentionally put the message in queue. And it works exactly as expected ! 
The message is queued and in the response tab, I can see the message I wrote (this why we have several attempt and why we have a Response Date).
Every 10 seconds the message is retried and as soon as the patient is creating, the message leave the queued and is treated successfully.
It's mandatory to preserve the order (by "admission" (aka "visitId")) by the way, that's why we didn't have enabled "Rotate".

What I expected is a maximum of 8 different "admission" with a "Response Date"/"Send attempts", not 3 :/
